### PR TITLE
VR-3587 commit with diff

### PR DESCRIPTION
**Pull request by evgeniilezhnin**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 1]
```diff
@@ -57,7 +62,7 @@ public String setBlobs(List<BlobContainer> blobContainers, FileHasher fileHasher
     }
   }
 
-  private Blob getBlob(Session session, InternalFolderElementEntity folderElementEntity)
+  public Blob getBlob(Session session, InternalFolderElementEntity folderElementEntity)
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: not really required?

**Comment by ravishetye** [comment]
Comment in Blob diff PR on the ordering of commits will affect this PR

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java (hunk 1 above)
Comment (ad-47) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: reverted

File: backend/src/main/java/ai/verta/modeldb/versioning/VersioningServiceImpl.java [hunk 2]
```diff
@@ -228,20 +228,38 @@ public void createCommit(
...
         throw new ModelDBException(
             "Parent commits not found in the CreateCommitRequest", Code.INVALID_ARGUMENT);
       }
-      List<BlobContainer> blobContainers = validateBlobs(request);
-      UserInfo currentLoginUserInfo = authService.getCurrentLoginUserInfo();
+      List<BlobContainer> blobContainers;
+
+      final RepositoryFunction repositoryFunction =
+          (session) -> repositoryDAO.getRepositoryById(session, request.getRepositoryId());
+      if (request.getBlobsCount() != 0) {
```
[the remark this answers is not in the document]
Comment (ad-47) [reply backend/src/main/java/ai/verta/modeldb/versioning/VersioningServiceImpl.java]: added

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 3]
```diff
@@ -577,4 +585,333 @@ boolean childContains(Set<?> list, Set<?> sublist) {
...
+  @Override
+  public List<BlobContainer> convertBlobDiffsToBlobs(
+      CreateCommitRequest request,
+      RepositoryFunction repositoryFunction,
+      CommitFunction commitFunction)
+      throws ModelDBException {
+    try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+      RepositoryEntity repositoryEntity = repositoryFunction.apply(session);
+      CommitEntity commitEntity = commitFunction.apply(session, session1 -> repositoryEntity);
+      Set<BlobExpanded> locationBlobListCommit =
```
[the remark this answers is not in the document]
Comment (ad-47) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: changed

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 4]
```diff
@@ -577,4 +585,333 @@ boolean childContains(Set<?> list, Set<?> sublist) {
...
+          getCommitBlobList(session, commitEntity.getRootSha(), new ArrayList<>());
+      Map<String, BlobExpanded> locationBlobsMapCommit =
```
[the remark this answers is not in the document]
Comment (ad-47) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: changed

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 5]
```diff
@@ -577,4 +585,333 @@ boolean childContains(Set<?> list, Set<?> sublist) {
...
+        // Apply the diffs on top of commit_base
+        if (blobDiff.getStatus() == DiffStatus.ADDED) {
+          // If a blob was added in the diff, add it on top of commit_base (doesn't matter
+          // if it was present already or not)
+          BlobExpanded blobExpanded = convertBlobDiffToBlobExpand(blobDiff);
+          blobContainers.add(blobExpanded);
+        } else {
+          if (blobDiff.getStatus() == DiffStatus.DELETED) {
+            // If a blob was deleted, delete if from commit_base if present
+            locationBlobsMapCommit.remove(String.join("#", locationList));
```
[the remark this answers is not in the document]
Comment (ad-47) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: Added exception

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 6]
```diff
@@ -577,4 +585,333 @@ boolean childContains(Set<?> list, Set<?> sublist) {
...
+            "Invalid blob type found in BlobDiff", Status.Code.INVALID_ARGUMENT);
+    }
+  }
+
+  private boolean isAtomic(BlobDiff blobDiff) {
+    return blobDiff.getContentCase() == ContentCase.CODE
+        || blobDiff.getContentCase() == ContentCase.ENVIRONMENT;
+  }
+
+  private void checkType(BlobDiff blobDiff, BlobExpanded existingBlob) throws ModelDBException {
```
[the remark this answers is not in the document]
Comment (ad-47) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: added

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 7]
```diff
@@ -577,4 +585,333 @@ boolean childContains(Set<?> list, Set<?> sublist) {
...
+      case DATASET:
+      case CONFIG:
+      case CONTENT_NOT_SET:
+      default:
+        throw new ModelDBException(
+            "Invalid blob type found in BlobDiff", Status.Code.INVALID_ARGUMENT);
+    }
+  }
+
+  private boolean isAtomic(BlobDiff blobDiff) {
```
[the remark this answers is not in the document]
Comment (ad-47) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: changed

File: backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java [hunk 8]
```diff
@@ -577,4 +585,333 @@ boolean childContains(Set<?> list, Set<?> sublist) {
...
+    return blobDiff.getContentCase() == ContentCase.CODE
+        || blobDiff.getContentCase() == ContentCase.ENVIRONMENT;
+  }
+
+  private void checkType(BlobDiff blobDiff, BlobExpanded existingBlob) throws ModelDBException {
+    BlobExpanded blobDiffExpanded = convertBlobDiffToBlobExpand(blobDiff);
```
[the remark this answers is not in the document]
Comment (ad-47) [reply backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java]: changed

**Pull request merged**